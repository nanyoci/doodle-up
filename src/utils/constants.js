export const API_URL = 'https://doodle-up-backend.herokuapp.com';