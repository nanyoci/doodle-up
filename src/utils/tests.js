import React from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { render } from '@testing-library/react';

import rootReducer from '../redux/rootReducer';

export const renderWithRedux = (component, initialState = {}) => {
    const store = createStore(
        rootReducer,
        initialState,
        applyMiddleware(thunk),
    )

    const renderedComponent = <Provider store={store}>{component}</Provider>;

    return {
        ...render(renderedComponent),
        store
    };
}

export const renderWithRouter = (component, route = '/') => {
    const history = createMemoryHistory({ initialEntries: [route] });
    const renderedComponent = <Router history={history}>{component}</Router>;

    return {
        ...render(renderedComponent),
        history
    };
}

export const renderWithReduxRouter = (component, initialState = {}, route = '/') => {
    const store = createStore(
        rootReducer,
        initialState,
        applyMiddleware(thunk),
    )
    const history = createMemoryHistory({ initialEntries: [route] });

    const renderedComponent = (
        <Provider store={store}>
            <Router history={history}>{component}</Router>
        </Provider>
    );

    return {
        ...render(renderedComponent),
        history,
        store
    };
}