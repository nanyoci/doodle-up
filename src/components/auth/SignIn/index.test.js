import React from 'react'
import { cleanup, fireEvent, waitForElement, act } from '@testing-library/react'
import axiosMock from 'axios';

import LoginPage from './index.js';
import { renderWithReduxRouter } from './../../../utils/tests';

jest.mock('axios');

beforeEach(() => {
    window.HTMLMediaElement.prototype.load = () => { /* do nothing */ };
    window.HTMLMediaElement.prototype.play = () => { /* do nothing */ };
    window.HTMLMediaElement.prototype.pause = () => { /* do nothing */ };
    window.HTMLMediaElement.prototype.addTextTrack = () => { /* do nothing */ };
    localStorage.removeItem('username');
})

afterEach(() => {
    cleanup();
    jest.clearAllMocks();
});

it('should take a snapshot', async () => {
    let asFragment = {};
    await act(async () => { ({ asFragment } = renderWithReduxRouter(<LoginPage />)) });
    expect(asFragment()).toMatchSnapshot();
})

it('allows the user to login successfully', async () => {
    const dummyResponse = "DoodleUp"

    axiosMock.post.mockResolvedValueOnce({
        "data": dummyResponse
    });

    const { getByTestId, store } = renderWithReduxRouter(<LoginPage />);
    const emailField = await waitForElement(() => getByTestId("emailField"));
    const passwordField = await waitForElement(() => getByTestId("passwordField"));
    const loginButton = await waitForElement(() => getByTestId("loginButton"));

    await act(async () => {
        fireEvent.change(emailField, {
            target: { value: 'doodleup@test.com' },
        })
        fireEvent.change(passwordField, {
            target: { value: 'doodleup1234' },
        })
    });

    await act(async () => {
        fireEvent.click(loginButton)
    });

    // post request to backend is made
    expect(axiosMock.post).toHaveBeenCalledTimes(1)
    // error state is null
    expect(store.getState().authReducer.error).toEqual('')
    // no error message displayed
    expect(() => getByTestId("auth-error-signin")).toThrow('Unable to find an element by: [data-testid="auth-error-signin"]');
    expect(localStorage.getItem('username')).toEqual(dummyResponse);
})

it('reject user login with wrong password', async () => {
    axiosMock.post.mockRejectedValueOnce();
    const expected_error_message = "Email and password do not match"

    const { getByTestId, store } = renderWithReduxRouter(<LoginPage />);

    const emailField = await waitForElement(() => getByTestId("emailField"));
    const passwordField = await waitForElement(() => getByTestId("passwordField"));
    const loginButton = await waitForElement(() => getByTestId("loginButton"));

    await act(async () => {
        fireEvent.change(emailField, {
            target: { value: 'doodleup@test.com' },
        })
        fireEvent.change(passwordField, {
            target: { value: 'wrong password' },
        })
    });

    await act(async () => {
        fireEvent.click(loginButton)
    });

    // post request to backend for authentication
    expect(axiosMock.post).toHaveBeenCalledTimes(1)
    // error state is updated
    expect(store.getState().authReducer.error).toEqual(expected_error_message)
    // error message is displayed on the page
    const err = await waitForElement(() => getByTestId("auth-error-signin"));
    expect(err.textContent).toEqual(expected_error_message)
    // no username stored in localStorage
    expect(localStorage.getItem('username')).toEqual(null)
})