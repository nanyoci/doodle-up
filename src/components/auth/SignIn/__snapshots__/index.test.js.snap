// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="main-content"
  >
    <audio />
    <div
      class="wrapper"
    >
      <img
        alt="DoodleUp"
        class="logo-center"
        src="doodle-up.svg"
      />
      <img
        alt="Open book"
        class="book-corner"
        src="book.svg"
      />
      <img
        alt="Blue splat of paint"
        class="splat-blue-corner"
        src="splat-blue.png"
      />
      <img
        alt="Pink splat of paint"
        class="splat-pink-corner"
        src="splat-pink.png"
      />
      <img
        alt="Paint brush and palette"
        class="paint-corner"
        src="paint.svg"
      />
      <div
        class="auth-container"
      >
        <h1>
          Sign In
        </h1>
        <form
          class="auth-form"
        >
          <div
            class="fill-in-form"
          >
            <div
              class="fill-in"
            >
              <h2
                class="fill-name"
              >
                Email
              </h2>
              <input
                class="input"
                data-testid="emailField"
                name="email"
                placeholder="Email Address"
                type="text"
                value=""
              />
            </div>
            <div
              class="fill-in"
            >
              <h2
                class="fill-name"
              >
                Password
              </h2>
              <input
                class="input"
                data-testid="passwordField"
                name="password"
                placeholder="Password"
                type="password"
                value=""
              />
            </div>
            <div
              class="menu-buttons"
            />
            <div
              class=" menu-buttons"
            >
              <button
                class="btn btn-primary"
                data-testid="loginButton"
                id="loginButton"
                type="submit"
              >
                Sign In
              </button>
            </div>
          </div>
        </form>
        <br />
        <a
          class="auth-link"
          href="/signup"
        >
          Create an account? Sign up
        </a>
      </div>
    </div>
    <div
      class="control-buttons"
    >
      <button
        class="ui yellow massive circular icon button"
        id="musicControlBtn"
      >
        <i
          aria-hidden="true"
          class="mute icon"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
